# Data Cleaning with Pandas

In this notebook we'll go through a few basic data cleaning steps that should be performed on all new datasets where necessary.

We'll go through the process with both the `orders` and `orderlines` datasets. You can then practice these skills by cleaning the `products` dataset yourself

In [1]:
import pandas as pd

In [2]:
# orders.csv
url = "https://drive.google.com/file/d/1Vu0q91qZw6lqhIqbjoXYvYAQTmVHh6uZ/view?usp=sharing" 
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orders = pd.read_csv(path)

# orderlines.csv
url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing" 
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orderlines = pd.read_csv(path)

One of the best ways to begin data cleaning is by exploring using `.info()`. This will tell us:
* The shape of the DataFrame
* The names of the columns
* If there are any missing values
* The datatypes of the columns

By exploring the missing values and correcting any incorrect datatypes, we often come across inconsistencies in our data.

Beyond this, we should also have a **check for any duplicate rows**. 

Let's first deal with the duplicates, as it's nice and easy, then we'll explore what `.info()` has to tell us.

In [3]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


In [4]:
orderlines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


## 1.&nbsp; Duplicates
We can check for duplicates using the pandas [.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) method. 

We can then delete these rows, if we wish, using [.drop_duplicates()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html)

In [6]:
# orders
orders.duplicated().sum()

0

In [7]:
# orderlines
orderlines.duplicated().sum()

0

We have no duplicate rows in either DataFrame. Easy, there is no problem to solve. Normally though, if there were some duplicates, we'd drop the extra rows.

In [8]:
names = ["Ana", "Erika", "Ana", "Javi", "Maria", "Maria", "Ana"]
ages = [29, 22, 29, 50, 23, 23, 29]




people = pd.DataFrame({"name":names,
                       "age":ages
                      })

people

name  age
0    Ana   29
1  Erika   22
2    Ana   29
3   Javi   50
4  Maria   23
5  Maria   23
6    Ana   29

In [9]:
people.duplicated()

0    False
1    False
2     True
3    False
4    False
5     True
6     True
dtype: bool

In [10]:
people.duplicated().sum()

3

In [13]:
people.drop_duplicates(inplace=True)

In [14]:
people

name  age
0    Ana   29
1  Erika   22
3   Javi   50
4  Maria   23

# 2.&nbsp; `.info()`

In [ ]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


* `total_paid` has 5 missing values
* `created_date` should become datetime datatype

In [ ]:
orderlines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


* `date` should be a datetime datatype
* `unit_price` should be a float datatype

## 3.&nbsp; Missing values

### 3.1.&nbsp; Orders
* `total_paid` has 5 missing values

In [16]:
orders.isna().sum()

order_id        0
created_date    0
total_paid      5
state           0
dtype: int64

In [17]:
orders.total_paid.isna().sum()

5

In [20]:
orders.shape[0]

226909

In [23]:
round((orders.total_paid.isna().sum()/orders.shape[0])*100, 5)

0.0022

In [ ]:
print(f"5 missing values represents {((orders.total_paid.isna().sum() / orders.shape[0])*100).round(5)}% of the rows in our DataFrame")

5 missing values represents 0.0022% of the rows in our DataFrame


As there is such a tiny amount of missing values, we will simply delete these rows, as we have enough data without them.

In [25]:
orders.dropna(inplace=True)

In [26]:
orders.isna().sum()

order_id        0
created_date    0
total_paid      0
state           0
dtype: int64

In [ ]:
orders = orders.loc[~orders.total_paid.isna(), :]# not

Should you have a significant number of missing values in the future, you have a choice: 
+ you can impute the values
+ you can take the values from other DataFrames, if they are present there
+ you can delete the values
+ or any number of other creative solutions

Please, always consider how much time you have on your project, and what impact your method of choice will have on your final assesment.

### 3.2.&nbsp; Orderlines
There are no missing values in `orderlines`

## 4.&nbsp; Datatypes

### 4.1.&nbsp; Orders
* `created_date` should become datetime datatype

In [27]:
orders["created_date"] = pd.to_datetime(orders["created_date"])

In [28]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226904 non-null  int64         
 1   created_date  226904 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226904 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 8.7+ MB


### 4.1.&nbsp; Orderlines
* `date` should be a datetime datatype
* `unit_price` should be a float datatype

#### 4.1.1.&nbsp; `date`

In [ ]:
orderlines["date"] = pd.to_datetime(orderlines["date"])

#### 4.1.2.&nbsp;`unit_price`

In [30]:
orderlines["unit_price"] = pd.to_numeric(orderlines["unit_price"])
print("hello")

ValueError: Unable to parse string "1.137.99" at position 6

As you can see when we try to convert `unit_price` to a numerical datatype, we receive a `ValueError` telling us that pandas doesn't understand the number `1.137.99`. This is probably because numbers cannot have 2 decimal points. Let's see if there are any other numbers like this.

In [32]:
orderlines.unit_price.str.contains("\d+\.\d+\.\d+").value_counts(normalize=True)

False    0.876969
True     0.123031
Name: unit_price, dtype: float64

Looks like over 36000 rows in `orderlines` are affected by this problem. Let's work out how much that is as a percentage of our total data.

In [ ]:
two_dot_percentage = ((orderlines.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()[1] / orderlines.shape[0])*100).round(2)
print(f"The 2 dot problem represents {two_dot_percentage}% of the rows in our DataFrame")

The 2 dot problem represents 12.3% of the rows in our DataFrame


This is a bit of a tricky decision as 12.3% is a significant amount of our data... and we might even end up losing a larger portion of our data than this too. For the moment we will delete the rows as we only have 2 weeks for this project and I'd like some quick, accurate results to show. If we have time at the end, we can come back and investigate this problem further, maybe there's a solution?

Each row of `orderlines` represents a product in an order. For example, if order number 175 contained 3 seperate products, then order 175 would have 3 rows in `orderlines`, one row for each of the products. If 2 of those products have 'normal' prices (14.99, 15.85) and 1 has a price with 2 decimal points (1.137.99), we need to remove the whole order and not just the affected row. If we only remove the row with 2 decimal places then any later analysis about products and prices could be misleading.

We therefore need to find the order numbers associated with the rows that have 2 decimal points, and then remove all the associated rows.

In [36]:
orderlines[orderlines.unit_price.str.contains("\d+\.\d+\.\d+") == True].index

Int64Index([     6,     11,     15,     43,     59,     61,     64,     67,
                69,     82,
            ...
            293728, 293748, 293756, 293790, 293838, 293862, 293887, 293889,
            293911, 293936],
           dtype='int64', length=36169)

In [ ]:
#10002

In [38]:
orderlines.drop(index=orderlines[orderlines.unit_price.str.contains("\d+\.\d+\.\d+") == True].index, inplace=True)

In [39]:
orderlines

id  id_order  product_id  product_quantity      sku unit_price  \
0       1119109    299539           0                 1  OTT0133      18.99   
1       1119110    299540           0                 1  LGE0043     399.00   
2       1119111    299541           0                 1  PAR0071     474.05   
3       1119112    299542           0                 1  WDT0315      68.39   
4       1119113    299543           0                 1  JBL0104      23.74   
...         ...       ...         ...               ...      ...        ...   
293978  1650199    527398           0                 1  JBL0122      42.99   
293979  1650200    527399           0                 1  PAC0653     141.58   
293980  1650201    527400           0                 2  APP0698       9.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
0       2017-01-01 00:07:19  
1       2017-01-01 00:19:45  
2       2017-01-01 00:20:57  
3       2017-01-01 00:51:40  
4       2017-01-01 01:06:38  
...                     ...  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36  

[257814 rows x 7 columns]

In [ ]:
two_dot_order_ids_list = orderlines.loc[orderlines.unit_price.str.contains("\d+\.\d+\.\d+"), "id_order"]
orderlines = orderlines.loc[~orderlines.id_order.isin(two_dot_order_ids_list)]

In [ ]:
orderlines.shape[0]

216250

We still have 216250 rows in orderlines to work with. This should be more than enough for our evaluation.

Now that all of the 2 decimal point prices have been removed, let's try again to convert the column `unit_price` to the correct datatype.

In [40]:
orderlines["unit_price"] = pd.to_numeric(orderlines["unit_price"])

It worked perfectly

# Challenge: Clean the `products` DataFrame
Now it's your turn. Use the lessons you learnt above and clean the products DataFrame. You don't have to copy exactly what we did. Think about the consequences of your actions, sometimes it is ok to delete rows, other times you may wish to come up with more creative solutions.

In [41]:
# products.csv
url = "https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=sharing" 
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
products = pd.read_csv(path)

In [42]:
products.head()

sku                                           name  \
0  RAI0007              Silver Rain Design mStand Support   
1  APP0023              Apple Mac Keyboard Keypad Spanish   
2  APP0025               Mighty Mouse Apple Mouse for Mac   
3  APP0072  Apple Dock to USB Cable iPhone and iPod white   
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   

   in_stock      type  
0         1      8696  
1         0  13855401  
2         0      1387  
3         0      1230  
4         1      1364

We'll go through the steps above in order
* Duplicates
* Missing values
* Datatypes

But I think we can all see straight away from `products.head()` above that some of the prices in `promo_price` look wrong. Let's make sure we deal with this later.

In [43]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


## Duplicates

In [46]:
products.duplicated().sum()

0

Wow, that's a lot of duplicates. Let's get rid of them.

In [45]:
products.drop_duplicates(inplace=True)

## `.info()`

In [ ]:
products.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10580 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10580 non-null  object
 1   name         10580 non-null  object
 2   desc         10573 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10580 non-null  object
 5   in_stock     10580 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


In [47]:
products.isna().sum()

sku             0
name            0
desc            7
price          46
promo_price     0
in_stock        0
type           50
dtype: int64

In [55]:
#products.drop(columns=['type'])

In [49]:
products[products.desc.isna()]

sku                                               name desc  \
16126  WDT0211-A  Open - Purple 2TB WD 35 "PC Security Mac hard ...  NaN   
16128  APP1622-A  Open - Apple Smart Keyboard Pro Keyboard Folio...  NaN   
17843    PAC2334             Synology DS718 + NAS Server | 10GB RAM  NaN   
18152  KAN0034-A  Open - Kanex USB-C Gigabit Ethernet Adapter Ma...  NaN   
18490    HTE0025  Hyper Pearl 1600mAh battery Mini USB Mirror an...  NaN   
18612    OTT0200    OtterBox External Battery Power Pack 20000 mAHr  NaN   
18690  HOW0001-A  Open - Honeywell thermostat Lyric zonificador ...  NaN   

           price promo_price  in_stock      type  
16126        107     814.659         0      1298  
16128  1.568.206   1.568.206         0      1298  
17843     566.35   5.659.896         0  12175397  
18152      29.99     237.925         0      1298  
18490      24.99       22.99         1      1515  
18612      79.99       56.99         1      1515  
18690     199.99   1.441.174         0  11905404

In [51]:
products.loc[products.desc.isna(), 'name']

16126    Open - Purple 2TB WD 35 "PC Security Mac hard ...
16128    Open - Apple Smart Keyboard Pro Keyboard Folio...
17843               Synology DS718 + NAS Server | 10GB RAM
18152    Open - Kanex USB-C Gigabit Ethernet Adapter Ma...
18490    Hyper Pearl 1600mAh battery Mini USB Mirror an...
18612      OtterBox External Battery Power Pack 20000 mAHr
18690    Open - Honeywell thermostat Lyric zonificador ...
Name: name, dtype: object

In [52]:
products.loc[products.desc.isna(), 'desc'] = products.loc[products.desc.isna(), 'name']

In [53]:
products.desc.isna().sum()

0

### Missing values
We can see from `.info()` above that we have missing values in `desc` and `price`

#### `desc`

In [ ]:
products.desc.isna().sum()

7

7 is a very small number to have missing, let's have a closer look

In [ ]:
products.loc[products.desc.isna(), :]

sku                                               name desc  \
16126  WDT0211-A  Open - Purple 2TB WD 35 "PC Security Mac hard ...  NaN   
16128  APP1622-A  Open - Apple Smart Keyboard Pro Keyboard Folio...  NaN   
17843    PAC2334             Synology DS718 + NAS Server | 10GB RAM  NaN   
18152  KAN0034-A  Open - Kanex USB-C Gigabit Ethernet Adapter Ma...  NaN   
18490    HTE0025  Hyper Pearl 1600mAh battery Mini USB Mirror an...  NaN   
18612    OTT0200    OtterBox External Battery Power Pack 20000 mAHr  NaN   
18690  HOW0001-A  Open - Honeywell thermostat Lyric zonificador ...  NaN   

           price promo_price  in_stock      type  
16126        107     814.659         0      1298  
16128  1.568.206   1.568.206         0      1298  
17843     566.35   5.659.896         0  12175397  
18152      29.99     237.925         0      1298  
18490      24.99       22.99         1      1515  
18612      79.99       56.99         1      1515  
18690     199.99   1.441.174         0  11905404

We have 2 choices here:
* We can quickly and easily remove these rows. 
* Or, alternatively, the products names here are quite descriptive, so I'm tempted to just copy them to the description column, so that there is a description if we later want utilise this column. I wouldn't recommend this if this DataFrame was the source of truth for our website. But this is not the case here, and we're not faking any information (guessing a price or so), so I'm happy with this option

In [ ]:
products.loc[products.desc.isna(), "desc"] = products.loc[products.desc.isna(), "name"]

Did you also notice above that we have the dreaded two decimal point problem in both the `price` and `promo_price` columns? We can also see prices with 3 decimal places, prices should have 2 decimal places: this gives us more cause for concern

#### `price`

In [56]:
products.price.isna().sum()

46

In [57]:
print(f"The missing values in price are {((products.price.isna().sum() / products.shape[0]) * 100).round(2)}% of all rows in the DataFrame")

The missing values in price are 0.43% of all rows in the DataFrame


Let's simply delete these rows to ensure that we can trust the numbers in our final DataFrame. Afterall, the price is very important when investigating discounts.

In [58]:
products = products.loc[~products.price.isna(), :]

### Data types

We saw from looking at the output of `.info()` that both `price` and `promo_price` have been stored as objects and not as a numerical datatypes. We also saw while solving other problems that both columns have some prices with 3 decimal places and others with 2 decimal points - the latter will prevent us from converting the datatype to numerical, so first we must investigate and solve these problems.

#### `price`

First, let's see how many values are affected by the 2 problems mentioned above.

In [60]:
pd.to_numeric(products.price)

ValueError: Unable to parse string "1.639.792" at position 504

In [ ]:
price_problems_number = products.loc[(products.price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products.price.astype(str).str.contains("\d+\.\d{3,}")), :].shape[0]
price_problems_number

542

In [ ]:
print(f"The column price has in total {price_problems_number} wrong values. This is {round(((price_problems_number / products.shape[0]) * 100), 2)}% of the rows of the DataFrame")

The column price has in total 542 wrong values. This is 5.15% of the rows of the DataFrame


5.15% is a reasonable amount of our data. However, the price column will be important to understanding discounts, so I'd like it to be very trustworthy as we are basing business decisions on it. Therefore, we'll delete these rows

In [61]:
products = products.loc[(~products.price.astype(str).str.contains("\d+\.\d+\.\d+"))&(~products.price.astype(str).str.contains("\d+\.\d{3,}")), :]

To complete our task, let's convert the column to a numeric datatype

In [64]:
products.loc[:,"price"] = pd.to_numeric(products["price"])

C:\Users\dell\AppData\Local\Temp\ipykernel_16212\2353184102.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  products.loc[:,"price"] = pd.to_numeric(products["price"])


#### `promo_price`

Again, let's begin by seeing how many values are affected by the 2 problems

In [65]:
promo_problems_number = products.loc[(products.promo_price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products.promo_price.astype(str).str.contains("\d+\.\d{3,}")), :].shape[0]
promo_problems_number

9232

In [66]:
print(f"The column promo_price has in total {promo_problems_number} wrong values. This is {round(((promo_problems_number / products.shape[0]) * 100), 2)}% of the rows of the DataFrame")

The column promo_price has in total 9232 wrong values. This is 92.39% of the rows of the DataFrame


WOW!!! That's a lot of wrong data. Let's have a quick investigate to check that's correct. We'll make a DataFrame by copy-pasting the code we used above and then look at a large sample to check that all the numbers in the `promo_price` column really have either 2 decimal points or 3 decimal places.

In [67]:
promo_price_df = products.loc[(products.promo_price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products.promo_price.astype(str).str.contains("\d+\.\d{3,}")), :]
promo_price_df.sample(50)

sku                                               name  \
11691    LOG0199  Create Logitech QWERTY keyboard cover 129 Span...   
14876    STA0052                  Startech USB-C 4K HDMI 60Hz White   
11242    PAC1340       Pack QNAP TS-453A | 8GB RAM | WD 8TB Network   
18961    NET0012  Arlo Netgear Pro 2 Video Surveillance Security...   
11139    SPE0160  Speck SeeThru housing Macbook Retina 12 "Trans...   
18437    APP2668  Apple iMac Pro 27 "18-core Intel Xeon W 23GHz ...   
19321    BEL0376            Belkin Travel Support Apple Watch Black   
18638    PAC2484  Synology DS218 NAS Server | 16TB (2x8TB) Seaga...   
17192    AP20246  Like new - Apple MacBook Pro Retina 13 "Core i...   
12       APP0111                     Apple USB Ethernet Adapter Mac   
17579    APP2500  Apple Watch GPS 42mm Case Series 3 Aluminum Si...   
2920     TRA0038  Transcend JetDrive Lite 360 ​​256GB Macbook Pr...   
13651    GRT0345            Griffin Reveal Case Black iPhone 6 Plus   
2468     LIF0053  Lifeproof Fre Case Waterproof iPad Mini / Reti...   
15653    APP2001  Apple iPhone Silicone Case Cover 8 Plus / 7 Pl...   
17166    AP20241  Like new - Apple MacBook Pro 13 "Core i5 2GHz ...   
735      PAC0858  Apple Mac Pro 12-core 27 Ghz | 32GB RAM | 512G...   
17760    PAC2388  DS418play Synology NAS Server | 2GB RAM | 24TB...   
11166    PAC1324  Pack QNAP TS-253A | 8GB RAM | Seagate 8TB Iron...   
541      HGD0018       Henge Docks Docking Station MacBook Air 11 "   
18442    APP2673  Apple iMac Pro 27 "8-core Intel Xeon W 32GHz |...   
14864    NTE0110  NewerTech NuPower Charger 60W USB-C for MacBoo...   
14547    PAC1543  Crucial SSD expansion kit MX300 2TB RAM + 16GB...   
11684    LIF0107  LifeProof nüüd iPhone 6s Plus Waterproof Case ...   
16124  NTE0050-A          Open - NewerTech Complete set of 14 tools   
15128  ADN0039-A            Open - Adonit Jot Dash 2 Pointer Silver   
14403    APP1832  Apple MacBook Pro 13 "Core i5 2GHz | 8GB RAM |...   
14705    PAC1766                                   QNAP TS-431P NAS   
14612    MOP0100  Mophie Powerstation 6000mAh Battery Plus Space...   
11832    APP1509        Apple iPad Pro 12.9 "Wi-Fi 256GB Space Gray   
18770  APP2072-A  Open - Apple MacBook Pro 13 "Core i5 23GHz | 8...   
15404    REP0284                           iPhone 5c battery repair   
14841  WAC0084-A  Open - Wacom Cintiq 13HD interactive pen displ...   
13573    APP1736  42mm Apple Watch Series 2 Aluminum Case and Sp...   
18456    APP2687  Apple iMac Pro 27 "14-core Intel Xeon W 25GHz ...   
16415    APP2279  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
12735    PAC1471  OWC ThunderBay 4 Thunderbolt 2 | Seagate Deskt...   
18605    OTT0204      OtterBox USB Cable Twisted Lightning 3 meters   
11727    STM0091                 Dux STM iPad Mini Case rigid 4 Red   
68       MAK0009     Maclocks theft case iPad 2 3 and 4 transparent   
12240    PAC1396  (OEM) Pack Office Home & Student perpetual lic...   
18013    MOS0239         Moshi Vitros X iPhone Case Red Transparent   
12910    KEU0010  Airframe kenu Car Support iPhone 6 / 6S Plus W...   
1488     PAC0618           Pack Synology DS1515 + | WD 15TB Network   
14015    APP1873  Apple MacBook Pro 15 "Core i7 Touch Bar 29GHz ...   
13262    PUR0146  Puro TPU UltraSlim 03 "Case + Screen Protector...   
15940  TUC0304-A  Open - Tucano Nido Hard-Shell Case MacBook Pro...   
15547    SXA0036  One SecurityXtra Security Support Wall iPad Pr...   
16136    APP1451      Apple iPad Case Silicone Case Mini 4 Lavender   
11447  CRU0039-A                 (Open) Crucial 240GB SSD 7mm BX200   

                                                    desc     price  \
11691  Cover with backlit keyboard and smart plug for...    155.00   
14876  C USB-HDMI adapter compatible with Thunderbolt...     47.99   
11242  QNAP NAS TS-453A with 8GB of RAM Memory Hard +...   1096.59   
18961  Video surveillance camera system with 130-degr...   1049.99   
11139             Protective case for Mac

So we were correct, over 90% of the data in this column is corrupt. There's no point deleting all of these rows, then we would barely have a products table. Instead, as it's only this column that appears to be very untrustworthy, we will delete the column.

In [68]:
products_cl = products.drop(columns=["promo_price"])

In [69]:
products_cl

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price  in_stock  \
0           Aluminum support compatible with all MacBook  59.99         1   
1              USB ultrathin keyboard Apple Mac Spanish.  59.00         0   
2                                 mouse Apple USB cable.  59.00         0   
3                  IPhone dock and USB Cable Apple iPod.  25.00         0   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99         1   
...                                                  ...    ...       ...   
19321  compact and portable stand vertically or horiz...  29.99         1   
19322  Backpack with capacity of 14 liter compartment...  69.95         1   
19323  Backpack with capacity of 14 liter compartment...  69.95         1   
19324  Backpack with capacity of 14 liter compartment...  69.95         0   
19325  Backpack with capacity of 14 liter compartment...  69.95         1   

           type  
0          8696  
1      13855401  
2          1387  
3          1230  
4          1364  
...         ...  
19321     12282  
19322      1392  
19323      1392  
19324      1392  
19325      1392  

[9992 rows x 6 columns]

Obviously, there's now no need to convert `promo_price` to a numerical datatype

Don't forget to download/save your new DataFrames. Also, give them an obvious name, so that you know they are the cleaned version and not the original DataFrame.

In [ ]:
from google.colab import files

orders.to_csv("orders_cl.csv", index=False)
files.download("orders_cl.csv")

orderlines.to_csv("orderlines_cl.csv", index=False)
files.download("orderlines_cl.csv")

products_cl.to_csv("products_cl.csv", index=False)
files.download("products_cl.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>